### PR TITLE
Remove pyramid of doom and add tests 

### PR DESCRIPTION
**Pull request by mpark63**
# Refactoring... 
- replace .thens with async/await 
  - make sure to await .save(), and .exec() queries 
- remove mongoose errors 
- make import naming consistent 
- added some more error handling
  - check for missing request fields in all routes (`missingHandler`)
  - add some 404 checking 
- write tests for higher coverage 
  - runs alright individually; npm test though does not work 
- fixed bugs while testing 

### For Reviewers: 
Although the async await PR #68 had been approved already, kindly review my commits after [ef4be6b](https://github.com/uCredit-Dev/uCredit-API/pull/70/commits/ef4be6b0b4dbe72906eed2677830996ea583c437). 
the big ones are: 
- [update tests for higher coverage](https://github.com/uCredit-Dev/uCredit-API/pull/70/commits/a649a45ff580a546fe26c05305205b18eb91210f) 
- [edit routes while testing](https://github.com/uCredit-Dev/uCredit-API/pull/70/commits/2ffadac81dee0ee1c87626fb160591c14888fbaf)

### Testing?
- running npm test
- ran backend locally and played around. just for sanity checking. seems like no change made to functionality. 

### Nice to haves (that will probably be in a separate PR)
- tests for notification and comment routes :( 
- enable es lint and prettier
- error reporting compatible w/ render (logtail / papertrail?)

### CODE REVIEW

**Comment by mpark63** [comment]
Apologies for big PR. Had to write tests in same branch as async/await as I wanted to make sure no breaking changes were made while removing pyramids of doom! Kindly review.

File: data/updateSIS.js [hunk 1]
```diff
@@ -49,7 +49,7 @@ function addProperty(course, res) {
   course.tags =
     details.PosTags.length == 0 ? [] : getField(details.PosTags, "Tag");
   console.log("-------saving", course.number);
-  course.save();
+  await course.save();
```
Comment (mattliu-mygit) [inline data/updateSIS.js:52]: Are these awaits necessary?

File: routes/planReview.js [hunk 2]
```diff
@@ -19,62 +25,57 @@ router.post("/api/planReview/request", auth, async (req, res) => {
...
-    forbiddenHandler(res);
-  } else {
-    const review = await planReviews
+    return missingHandler(res, { plan_id, reviewer_id, reviewee_id });
+  }
+  if (req.user._id !== reviewee_id) {
+    return forbiddenHandler(res);
+  }
+  try {
+    const _review = await Reviews
```
Comment (mattliu-mygit) [inline routes/planReview.js:34]: Is this good naming convention?

Comment (mattliu-mygit) [reply routes/planReview.js:34]: yes? no? maybe so

File: routes/planReview.js [hunk 3]
```diff
@@ -19,62 +25,57 @@ router.post("/api/planReview/request", auth, async (req, res) => {
...
-          returnData(review, res);
-        })
-        .catch((err) => errorHandler(res, 500, err));
     }
+    const planReview = {
+      reviewee_id,
+      reviewer_id,
+      plan_id,
+      requestTime: Date.now(),
+      status: "PENDING",
```
Comment (mattliu-mygit) [inline routes/planReview.js:47]: Usually these constants should be put in an enum.
ie.
const PLAN_STATUS = {pending: "PENDING", review: "PLANREVIEW"}

File: routes/sso.js [hunk 4]
```diff
@@ -38,7 +38,7 @@ const samlStrategy = new saml.Strategy(
     issuer: SP_NAME,
     callbackUrl: `${BASE_URL}/api/login/callback`,
     decryptionPvk: PvK,
-    privateCert: PvK,
```
Comment (mattliu-mygit) [inline routes/sso.js:41]: pogger

File: routes/user.js [hunk 5]
```diff
@@ -25,64 +25,66 @@ router.get("/api/user", (req, res) => {
...
+      whitelisted_plan_ids: user.whitelisted_plan_ids,
+    }));
+    returnData(users, res);
+  } catch (err) {
+    errorHandler(res, 400, err);
+  }
 });
 
 // TODO: Temporarily disabled for OOSE dev deployment
 // if (DEBUG) {
```
Comment (mattliu-mygit) [inline routes/user.js:46]: We can probably uncomment this now... I don't see any big issues with uncommenting this and it's good.

File: tests/routes/SSO.spec.js [hunk 6]
```diff
@@ -1,35 +1,102 @@
 import mongoose from "mongoose";
 import supertest from "supertest";
 import createApp from "../../app";
-import User from "../../model/User";
+import Users from "../../model/User";
+import Sessions from "../../model/Session";
+import cryptoRandomString from "crypto-random-string";
+import { decodeToken } from "../../util/token";
+import { TEST_USER_1 } from "./testVars";
```
Comment (mattliu-mygit) [inline tests/routes/SSO.spec.js:9]: What is the statement and branch coverage?

**Review by mattliu-mygit — CHANGES_REQUESTED** [review]
Good job!!!! You blew me out of the water!!! This is amazing. Please note comments tho. :)

**Review by QinyuanWu — APPROVED** [review]
Great work Minseo!! These changes are much needed QAQ

**Comment by mpark63** [comment]
Screenshot 2023-01-08 at 2 12 41 PM
Screenshot 2023-01-08 at 2 12 53 PM

**Pull request merged**